### PR TITLE
Feature: homepage quick launch aka search

### PR DESCRIPTION
**Pull request by shamoon**
I made a thing! Its a "quick launcher"? aka global search of all services & bookmarks, triggered by just typing on the page. Its very simple but as a result its fast AF. Im sure I missed some things in here but overall pretty happy with this. Appreciate any feedback / tweaks. Video is all keyboard =)

Edit:
- A way to bring this up on mobile, or keep it desktop only?
- Should we search / include descriptions? 
  - Edit: added as [an option](https://github.com/benphelps/homepage/pull/409#issuecomment-1285970882)
- ~~Prioritization?~~ 
  - Edit: uses same order as services/bookmarks config, I think thats good
- ~~Highlighting?~~
  - Edit: Only for descriptions. [I dont like it for the main name.](https://user-images.githubusercontent.com/4887959/196809825-ceef596b-c5ff-4cc8-a3ba-93e26ccc322f.png)

https://user-images.githubusercontent.com/4887959/196631037-972fe92d-0ed1-44b1-bd0b-83ce615237ae.mov

Screen Shot 2022-10-19 at 8 19 29 PM

### CODE REVIEW

**Comment by nimodius** [comment]
This seems awesome, very useful! Cani test it somehow?

**Comment by shamoon** [comment]
> This seems awesome, very useful! Cani test it somehow?

Good question! I think you'd need to pull the repo but its pretty easy if you want to try.

Regarding including descriptions, curious on people's thoughts. I have an (uncommitted) change that:
- shows the description
- searches the description and shows highlighting only in the description to make it apparent why that result is included
- prioritizes name matches ahead of description matches

See screenshot, again, would love to hear what people think 🙂
Screen Shot 2022-10-19 at 3 25 07 PM

**Comment by Facinorous-420** [comment]
> > This seems awesome, very useful! Cani test it somehow?
> 
> Good question! I think you'd need to pull the repo but its pretty easy if you want to try.
> 
> Regarding including descriptions, curious on people's thoughts. I have an (uncommitted) change that:
> 
> * shows the description
> * searches the description and shows highlighting only in the description to make it apparent why that result is included
> * prioritizes name matches ahead of description matches
> 
> See screenshot, again, would love to hear what people think 🙂 Screen Shot 2022-10-19 at 3 25 07 PM

This looks really nice Im going to pull this branch and try it out!
I like it showing the description, but I don't know how I feel about it also searching the description.

**Comment by shamoon** [comment]
> This looks really nice Im going to pull this branch and try it out! I like it showing the description, but I don't know how I feel about it also searching the description.

Thanks! 😊 Yea I think I agree, I added them without searching, for now.

**Comment by nimodius** [comment]
I did try it and I think it works great, would love to have this. Once I used it a little I found myself missing it =)

Perhaps searching descriptions could be a setting, though only if many would use it (I dont think I would)

File: src/pages/index.jsx [hunk 1]
```diff
@@ -188,6 +191,28 @@ function Home({ initialSettings }) {
     }
   }, [i18n, settings, color, setColor, theme, setTheme]);
 
+  const [searching, setSearching] = useState(false);
+  const [searchString, setSearchString] = useState("");
+
+  useEffect(() => {
+    function handleKeyDown(e) {
+      if (e.target.tagName === "BODY") {
+        if (String.fromCharCode(e.keyCode).match(/(\w|\s)/g) && !(e. altKey || e.ctrlKey || e.metaKey || e.shiftKey)) {
```
Comment (JazzFisch) [inline src/pages/index.jsx:200]: It doesn't affect anything, but it's weird to see a space after e. and before altKey.

Comment (shamoon) [reply src/pages/index.jsx]: I guess lint doesnt mind that! fixed =)

**Comment by shamoon** [comment]
Okey-dokey I added an option for searching descriptions (defaults to off), e.g. 

```
quicklook:
    searchDescriptions: true
```

Screen Shot 2022-10-20 at 11 00 17 AM
vs
Screen Shot 2022-10-20 at 11 26 43 AM

and everybody wins =)

(note the first three items stay higher on the list because they matched the actual name)

**Comment by nimodius** [comment]
> Okey-dokey I added an option for searching descriptions (defaults to off), e.g.

Yes I think this is a good solution for all. Hope to have this feature soon very soon!

**Comment by moritz-john** [comment]
> quicklaunch:
>     searchDescriptions: true

Hello,
I think this new feature & configuration option(s) need to be added to the documentation or am I missing something?

image

**Pull request merged**